doc/neps/nep-0021-advanced-indexing.rst

```diff
@@ -11,30 +12,175 @@ Implementing intuitive and full featured advanced indexing
+multi-dimensional array data, so it is no surprise that many other libraries in
+the scientific Python ecosystem also support array indexing.
+
+Unfortunately, the full complexity of NumPy's indexing rules mean that it is
+both challenging and undesirable for other libraries to copy its behavior in all
+of its nuance. The only full implementation of NumPy-style indexing is NumPy
+itself. This includes projects like dask.array and h5py, which support *most*
+types of array indexing in some form, and otherwise attempt to copy NumPy's API
+exactly.
+
+Vectorized indexing in particular can be challenging to implement with array
+storage backends not based on NumPy. In contrast, indexing by 1D arrays along
+at least one dimension in the style of outer indexing is much more acheivable.
+This has led many libraries (including dask and h5py) to attempt to define a
+safe subset of NumPy-style indexing that is equivalent to outer indexing, e.g.,
+by only allowing indexing with an array along at most one dimension. However,
+this is quite challenging to do correctly in a general enough way to be useful.
+For example, the current versions of dask and h5py both handle mixed indexing
+in case 3 above inconsistently with NumPy. This is quite likely to lead to
+bugs.
+
+These inconsistencies, in addition to the broader challenge of implementing
+every type of indexing logic, make it challenging to write high-level array
+libraries like xarray or dask.array that can interchangeably index many types of
+array storage. In contrast, explicit APIs for outer and vectorized indexing in
+NumPy would provide a model that external libraries could reliably emulate, even
+if they don't support every type of indexing.
+
+High level changes
+------------------
+
+Inspired by multiple "indexer" attributes for controlling different types
+of indexing behavior in pandas, we propose to:
+
+1. Introduce ``arr.oindex[indices]`` which allows array indices, but
    uses outer indexing logic.
 2. Introduce ``arr.vindex[indices]`` which use the current
    "vectorized"/broadcasted logic but with two differences from
-   fancy indexing:
+   legacy indexing:
        
-   1. Boolean indices always use the outer indexing logic.
-      (Multi dimensional booleans should be allowed).
-   2. The integer index result dimensions are always the first axes
-      of the result array. No transpose is done, even for a single
-      integer array index.
-
-3. Plain indexing on the array will only give warnings and eventually
-   errors either:
-     
-   * when there is ambiguity between legacy fancy and outer indexing
-     (note that ``arr[[1, 2], :, 0]`` is such a case, an integer
-     can be the "second" integer index array),
-   * when any integer index array is present (possibly additional for
-     more then one boolean index array).
+   * Boolean indices are not supported. All indices must be integers,
+     integer arrays or slices.
```

Comment:
Probably worth noting why this is a good idea, and how the desired behavior can be built from `np.nonzero`

Reply:
Nevermind, I see this is below